### PR TITLE
Add `--fix-rng-seed` for reproducibility

### LINKED ISSUE #977: Show random seed for experiments.
In this [slurm ci](https://github.com/climate-machine/CLIMA/pull/938#issuecomment-615415685) Held-Suarez failed. I believe it was due to the random numbers being generated but I have no way to check (it passed locally and matched master when I forced the random seed).

I would recommend that any experiments that require random numbers display the seed they use.

For MPI jobs we need to display each ranks random seed.

This is related to closed issue #792

### PR DESCRIPTION
**Pull request by kpamnany**
# Description

Closes #977. Will update documentation in #1091.

I have

- [X] Written and run all necessary tests with CLIMA by including `tests/runtests.jl`
- [X] Followed all necessary [style guidelines](https://CliMA.github.io/CLIMA/latest/CodingConventions.html) and run `julia .dev/climaformat.jl .`
- [ ] Updated the documentation to reflect changes from this PR.

# For review by CLIMA developers

- [ ] There are no open pull requests for this already
- [ ] CLIMA developers with relevant expertise have been assigned to review this submission
- [ ] The code conforms to the [style guidelines](https://CliMA.github.io/CLIMA/latest/CodingConventions.html) and has consistent naming conventions. `julia .dev/format.jl` has been run in a separate commit.
- [ ] This code does what it is technically intended to do (all numerics make sense physically and/or computationally)

### CODE REVIEW

**Comment by kpamnany** [comment]
Do we want to add _use_ of this parameter in `slurmci-test.toml` or apply it to all tests in SlurmCI scripts? I'm inclined to the latter, but...

File: src/Driver/Driver.jl [hunk 1]
```diff
@@ -145,6 +149,9 @@ function parse_commandline(custom_settings)
         metavar = "<number>"
         arg_type = Int
         default = -1
+        "--fix-rng-seed"
+        help = "set RNG seed to a fixed value for reproducibility"
+        action = :store_true
```
Comment (charleskawczynski) [inline src/Driver/Driver.jl:154]:
```suggestion
        default = true
```
?

Comment (kpamnany) [reply src/Driver/Driver.jl:154]: It's a no-argument... argument. That is, you just add `--fix-rng-seed`, not `--fix-rng-seed=10` or anything. So when the argument is found, this action stores `true` in the relevant `Dict` entry. See [here](https://carlobaldassi.github.io/ArgParse.jl/stable/arg_table/#Available-actions-and-nargs-values-1) if you want to know the details!

File: src/Driver/Driver.jl [hunk 2]
```diff
@@ -253,6 +260,15 @@ function init(; disable_gpu = false, arg_settings = nothing)
     _init_array(atyp)
     Settings.array_type = atyp
 
+    # fix the RNG seeds if requested
+    if parsed_args["fix-rng-seed"]
+        rank = MPI.Comm_rank(MPI.COMM_WORLD)
+        for tid in 1:nthreads()
+            s = parse(Int, @sprintf("%d%03d", rank, tid))
```
Comment (simonbyrne) [inline src/Driver/Driver.jl:267]:
```suggestion
            s = 1000*rank + tid
```

**Comment by kpamnany** [comment]
I've added `--fix-rng-seed` to all tests in [`slurmci-test.toml`](https://github.com/CliMA/ClimateMachine.jl/commit/24125b51188cda3f4a5acfc494fe90b03957baa0) that use `rand()`. I'm guessing it isn't needed for all these. Please let me know from where it can/should be removed?

Attn: @jkozdon @lcw

**Comment by jkozdon** [comment]
Should we remove the seed from the tests / problems that set them in the drivers?

- `experiments/AtmosLES/bomex.jl`
- `tutorials/Atmos/dry_rayleigh_benard.jl`
- `test/Numerics/Mesh/mpi_getpartition.jl`
- `test/Numerics/LinearSolvers/bandedsystem.jl`
- `test/Numerics/LinearSolvers/cg.jl`
- `test/Numerics/LinearSolvers/bgmres.jl`
- `test/Numerics/Mesh/interpolation.jl`
- `test/Common/MoistThermodynamics/runtests.jl`
- `tutorials/Numerics/LinearSolvers/bgmres.jl`
- `test/Numerics/DGmethods/advection_diffusion/direction_splitting_advection_diffusion.jl`
- `test/Numerics/LinearSolvers/iterativesolvers.jl`
- `test/Numerics/LinearSolvers/columnwiselu.jl`
- `test/Numerics/DGmethods/conservation/sphere.jl`
- `tutorials/Atmos/risingbubble.jl`
- `tutorials/Numerics/LinearSolvers/cg.jl`

**Comment by simonbyrne** [comment]
> Should we remove the seed from the tests / problems that set them in the drivers?

We should do it for all the experiments and tutorials at least.

**Comment by kpamnany** [comment]
bors try

**Comment by kpamnany** [comment]
Clarifying the purpose of this argument according to my understanding:

- For non-test runs it is useful to have a non-fixed RNG seed. For such, this argument is useful because we can exclude the command line argument and get actual random numbers. We don't want to use an explicitly seeded RNG for these.
- For test runs, we want to use a fixed seed for repeatability. Using an explicitly seeded RNG, as is done in some of the tests, is fine, unless:
  - `rand()` calls may be made in parallel, which could happen when `init_state!()` runs in parallel, or
  - it is useful to run the test with actual random numbers.

So with this in mind, I've adjusted use of this argument. It is applied _only_ to tests in `slurmci-test.toml` that meet the above criteria. As far as I can tell, nothing that's invoked from `test/runtests.jl` is using `rand()` without explicitly seeding so this argument is _not_ used for this.

Tests that have this argument applied and have explicit seeding removed (if any) are:

- `test/Numerics/DGmethods/advection_diffusion/direction_splitting_advection_diffusion.jl`
- `test/Numerics/DGmethods/conservation/sphere.jl`
- `tutorials/Atmos/dry_rayleigh_benard.jl`
- `experiments/AtmosLES/bomex.jl`
- `experiments/AtmosGCM/heldsuarez.jl`

Tests which use `rand()` and the RNG is seeded explicitly (i.e. test results should be reproducible):

- `test/Common/MoistThermodynamics/runtests.jl`
- `test/Numerics/LinearSolvers/bgmres.jl`
- `test/Numerics/LinearSolvers/cg.jl`
- `test/Numerics/LinearSolvers/iterativesolvers.jl`
- `test/Numerics/LinearSolvers/columnwiselu.jl`
- `test/Numerics/Mesh/mpi_getpartition.jl`
- `tutorials/Numerics/LinearSolvers/bgmres.jl`
- `tutorials/Numerics/LinearSolvers/cg.jl`

Apart from these, the following do not actually use `rand()`:

- `test/Numerics/Mesh/interpolation.jl`
- `tutorials/Atmos/risingbubble.jl`
- `experiments/AtmosLES/taylor-green.jl`

**Comment by jkozdon** [comment]
> Clarifying the purpose of this argument according to my understanding:

Sounds good to me. I just want to CI to be reproducible so that when we have problems we can track them down.

Thanks for taking this on.

**Comment by kpamnany** [comment]
@akshaysridhar : need your help with this one. BOMEX reliably fails its test when run with `--fix-rng-seed` from this PR branch:
```
Test Failed at /central/home/kpamnany/ClimateMachine.jl-alt/experiments/AtmosLES/bomex.jl:524
  Expression: isapprox(result, FT(1); atol=0.004)
   Evaluated: isapprox(1.0040551f0, 1.0f0; atol=0.004)
```
The only change is that the RNG seed, that was previously set in BOMEX to 0, is now set to 1001 externally.

Any suggestions?

**Comment by akshaysridhar** [comment]
bors try

**Pull request merged**